### DISCUSSION TITLE
Include MSAL authentication

### DISCUSSION THREAD

**Discussion by SebastianSchuetze** [DarqueWarrior/vsteam#454] [Ideas]
Hi,

I wanted to discuss if we can use MSAL auth. [AndrewZtrhgf ](https://twitter.com/AndrewZtrhgf) on twitter contacted me to suggest to add MSAL auth to the module. What do you think @DarqueWarrior?

His example blog post:
https://doitpsway.com/how-to-get-all-azure-devops-organizations-in-your-azuread-directory-using-powershell

I wanna see if that makes sense security and productivity wise. Because it might be useful to authenticate as well initially over some flow and have maybe the module after getting the accessToken to generate a PAT automatically saved within a profile.

Could make the initial login easier. I haven't checked what is needed, but it may be a good addition.

cc: @ztrhgf

**Comment by ztrhgf** [comment]
From my point of view, PAT cannot be used to gather all DevOps organizations. Instead, you have to use MSAL (or oAuth probably) which was exactly my case, because I was working on some auditing solution that could help us to get an overview of every organization (and users, groups used there) we have attached to our AAD.

So by adding MSAL support, you could also add a new function to list all organizations (see the blog post how to do it) :)

**Comment by SebastianSchuetze** [comment]
I just talked with Andrew a bit more and the following URL seems not to be accessible via a PAT and only with an accessToken:

```
https://aexprodweu1.vsaex.visualstudio.com/_apis/EnterpriseCatalog/Organizations?tenantId=<tenantId> 
```
![image](https://user-images.githubusercontent.com/7256168/155332243-bdff45d3-f931-4d22-81bc-11858a51049f.png)
